### PR TITLE
Add fast_roc_auc

### PR DESCRIPTION
**Pull request by Innixma**
Related to #20 

Add `fast_roc_auc`, which uses a C++ compiled implementation of `roc_auc` for a >2x speedup.

This is an improvement over our prior `roc_auc` implementation (`ag_roc_auc`).

To use `fast_roc_auc`, you will need to compile the C++ code via the newly added `compile.sh`. If you attempt to run the code without doing so, it will error with a message containing the terminal command needed to compile the code.

Here is an example output from `run_bench_roc_auc.py`

```
Benchmarking roc_auc... (num_samples=2, num_repeats=1000
	Time = 0.3948 ms	| Rel Speedup = 1.0x	| Score = 0.0	| sk_roc_auc
	Time = 0.0538 ms	| Rel Speedup = 7.3x	| Score = 0.0	| ag_roc_auc
	Time = 0.0084 ms	| Rel Speedup = 46.8x	| Score = 0.0	| fast_roc_auc_cpp
Benchmarking roc_auc... (num_samples=10, num_repeats=1000
	Time = 0.4101 ms	| Rel Speedup = 1.0x	| Score = 0.1875	| sk_roc_auc
	Time = 0.0539 ms	| Rel Speedup = 7.6x	| Score = 0.1875	| ag_roc_auc
	Time = 0.0084 ms	| Rel Speedup = 48.7x	| Score = 0.1875	| fast_roc_auc_cpp
Benchmarking roc_auc... (num_samples=100, num_repeats=1000
	Time = 0.4213 ms	| Rel Speedup = 1.0x	| Score = 0.4277597402597403	| sk_roc_auc
	Time = 0.0559 ms	| Rel Speedup = 7.5x	| Score = 0.42775974025974023	| ag_roc_auc
	Time = 0.0092 ms	| Rel Speedup = 45.7x	| Score = 0.4277597402597403	| fast_roc_auc_cpp
Benchmarking roc_auc... (num_samples=1000, num_repeats=1000
	Time = 0.5901 ms	| Rel Speedup = 1.0x	| Score = 0.4813428059395802	| sk_roc_auc
	Time = 0.0900 ms	| Rel Speedup = 6.6x	| Score = 0.4813428059395801	| ag_roc_auc
	Time = 0.0204 ms	| Rel Speedup = 28.9x	| Score = 0.4813428059395801	| fast_roc_auc_cpp
Benchmarking roc_auc... (num_samples=2000, num_repeats=1000
	Time = 0.7811 ms	| Rel Speedup = 1.0x	| Score = 0.5006057186644378	| sk_roc_auc
	Time = 0.1594 ms	| Rel Speedup = 4.9x	| Score = 0.5006057186644378	| ag_roc_auc
	Time = 0.0585 ms	| Rel Speedup = 13.4x	| Score = 0.5006057186644378	| fast_roc_auc_cpp
Benchmarking roc_auc... (num_samples=5000, num_repeats=100
	Time = 1.3484 ms	| Rel Speedup = 1.0x	| Score = 0.4909501567205458	| sk_roc_auc
	Time = 0.4038 ms	| Rel Speedup = 3.3x	| Score = 0.49095015672054576	| ag_roc_auc
	Time = 0.2453 ms	| Rel Speedup = 5.5x	| Score = 0.49095015672054576	| fast_roc_auc_cpp
Benchmarking roc_auc... (num_samples=10000, num_repeats=100
	Time = 2.3446 ms	| Rel Speedup = 1.0x	| Score = 0.509796231110791	| sk_roc_auc
	Time = 0.8498 ms	| Rel Speedup = 2.8x	| Score = 0.5097962311107911	| ag_roc_auc
	Time = 0.5501 ms	| Rel Speedup = 4.3x	| Score = 0.509796231110791	| fast_roc_auc_cpp
Benchmarking roc_auc... (num_samples=100000, num_repeats=20
	Time = 23.7546 ms	| Rel Speedup = 1.0x	| Score = 0.4982425920099391	| sk_roc_auc
	Time = 12.0591 ms	| Rel Speedup = 2.0x	| Score = 0.4982425920099391	| ag_roc_auc
	Time = 6.8111 ms	| Rel Speedup = 3.5x	| Score = 0.4982425920099391	| fast_roc_auc_cpp
Benchmarking roc_auc... (num_samples=1000000, num_repeats=3
	Time = 280.7703 ms	| Rel Speedup = 1.0x	| Score = 0.4997539871865995	| sk_roc_auc
	Time = 154.8101 ms	| Rel Speedup = 1.8x	| Score = 0.4997539871865995	| ag_roc_auc
	Time = 82.6302 ms	| Rel Speedup = 3.4x	| Score = 0.4997539871865996	| fast_roc_auc_cpp
```

In terms of simulation speed, when combined with the `log_loss` optimization from #23, we are ~3x faster:

Prior to metric optimizations:
```
Fitting Fold 1/5 (R1S1)...
    train_datasets: 104/104 | train_tasks: 1040
    test_datasets : 26/26 | test_tasks : 260
    n_configs=608/608
2023-05-03 02:47:59,089	INFO worker.py:1538 -- Started a local Ray instance.
1	: Train: 3.17 | 17.89s	| Tr Change: 2.840	| Conf Options: 608	| ray | CatBoost_r9_BAG_L1
2	: Train: 2.7 | 135.53s	| Tr Change: 0.471	| Conf Options: 607	| ray | NeuralNetFastAI_r35_BAG_L1
3	: Train: 2.4 | 193.36s	| Tr Change: 0.292	| Conf Options: 606	| ray | LightGBM_r12_BAG_L1
4	: Train: 2.28 | 250.21s	| Tr Change: 0.127	| Conf Options: 605	| ray | ExtraTrees_r23_BAG_L1
5	: Train: 2.17 | 306.84s	| Tr Change: 0.108	| Conf Options: 604	| ray | LightGBM_r146_BAG_L1
6	: Train: 2.1 | 363.53s	| Tr Change: 0.073	| Conf Options: 603	| ray | CatBoost_r12_BAG_L1
7	: Train: 2.05 | 421.3s	| Tr Change: 0.047	| Conf Options: 602	| ray | CatBoost_r72_BAG_L1
8	: Train: 2.03 | 477.13s	| Tr Change: 0.024	| Conf Options: 601	| ray | LightGBM_r66_BAG_L1
9	: Train: 2.01 | 535.3s	| Tr Change: 0.018	| Conf Options: 600	| ray | ExtraTrees_r11_BAG_L1
10	: Train: 1.99 | 591.18s	| Tr Change: 0.016	| Conf Options: 599	| ray | RandomForest_r37_BAG_L1
selected ['CatBoost_r9_BAG_L1', 'NeuralNetFastAI_r35_BAG_L1', 'LightGBM_r12_BAG_L1', 'ExtraTrees_r23_BAG_L1', 'LightGBM_r146_BAG_L1', 'CatBoost_r12_BAG_L1', 'CatBoost_r72_BAG_L1', 'LightGBM_r66_BAG_L1', 'ExtraTrees_r11_BAG_L1', 'RandomForest_r37_BAG_L1']
test_score: 2.4736221676682826
```

This PR (Slight train/test score difference due to small rounding errors changing how ties occur slightly, can be ignored):
```
Fitting Fold 1/5 (R1S1)...
    train_datasets: 104/104 | train_tasks: 1040
    test_datasets : 26/26 | test_tasks : 260
    n_configs=608/608
2023-05-03 07:52:31,321	INFO worker.py:1538 -- Started a local Ray instance.
1	: Train: 3.17 | 12.87s	| Tr Change: 2.841	| Conf Options: 608	| ray | CatBoost_r9_BAG_L1
2	: Train: 2.69 | 58.34s	| Tr Change: 0.471	| Conf Options: 607	| ray | NeuralNetFastAI_r35_BAG_L1
3	: Train: 2.4 | 79.4s	| Tr Change: 0.290	| Conf Options: 606	| ray | LightGBM_r12_BAG_L1
4	: Train: 2.28 | 100.5s	| Tr Change: 0.128	| Conf Options: 605	| ray | ExtraTrees_r23_BAG_L1
5	: Train: 2.17 | 120.45s	| Tr Change: 0.110	| Conf Options: 604	| ray | LightGBM_r146_BAG_L1
6	: Train: 2.09 | 141.09s	| Tr Change: 0.073	| Conf Options: 603	| ray | CatBoost_r12_BAG_L1
7	: Train: 2.05 | 161.86s	| Tr Change: 0.047	| Conf Options: 602	| ray | CatBoost_r72_BAG_L1
8	: Train: 2.02 | 181.43s	| Tr Change: 0.024	| Conf Options: 601	| ray | LightGBM_r82_BAG_L1
9	: Train: 2.0 | 201.27s	| Tr Change: 0.019	| Conf Options: 600	| ray | RandomForest_r16_BAG_L1
10	: Train: 1.99 | 221.73s	| Tr Change: 0.019	| Conf Options: 599	| ray | NeuralNetFastAI_r41_BAG_L1
selected ['CatBoost_r9_BAG_L1', 'NeuralNetFastAI_r35_BAG_L1', 'LightGBM_r12_BAG_L1', 'ExtraTrees_r23_BAG_L1', 'LightGBM_r146_BAG_L1', 'CatBoost_r12_BAG_L1', 'CatBoost_r72_BAG_L1', 'LightGBM_r82_BAG_L1', 'RandomForest_r16_BAG_L1', 'NeuralNetFastAI_r41_BAG_L1']
test_score: 2.3743008798881537
```

### CODE REVIEW

**Comment by Innixma** [comment]
Thanks to @diditforlulz273 for the very useful [fastauc](https://github.com/diditforlulz273/fastauc) repo that inspired this optimization!

**Comment by diditforlulz273** [comment]
happy to serve! tag me somewhere when this repo goes big lol

I actually created this to speed up bootstrapping AUC difference on models predictions, works like a charm compared to the standard sklearn version

File: autogluon_zeroshot/metrics/_roc_auc_cpp/__init__.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    def __init__(self):
+        try:
+            self._handle = ctypes.CDLL(os.path.dirname(os.path.realpath(__file__)) + os.path.sep + "cpp_auc.so")
+        except OSError:
+            command_to_compile = f"cd {os.path.dirname(os.path.realpath(__file__))} && .{os.path.sep}compile.sh"
+            raise OSError(f'Missing cpp_auc.so compiled file... '
+                          f'You must first compile the C++ code to use this metric. '
+                          f'Run the below terminal command to compile the code:\n'
+                          f'{command_to_compile}')
```
Comment (geoalgo) [inline autogluon_zeroshot/metrics/_roc_auc_cpp/__init__.py:20]: Given that we are just checking the file presence, it may be better to just to that directly to avoid catching errors not mentioned in the message:
```suggestion
        plugin_path = Path(__file__) / "cpp_auc.so"
        command_to_compile = f"cd {Path(__file__)} && .{os.path.sep}compile.sh"
        assert pluging_path.exists(), f'Missing cpp_auc.so compiled file... '
                          f'You must first compile the C++ code to use this metric. '
                          f'Run the below terminal command to compile the code:\n'
                          f'{command_to_compile}'

        self._handle = ctypes.CDLL(plugin_path)       
```

**Review by geoalgo — APPROVED** [review]
LGTM, cool stuff!

File: autogluon_zeroshot/metrics/_roc_auc_cpp/__init__.py (hunk 1 above)
Comment (geoalgo) [reply autogluon_zeroshot/metrics/_roc_auc_cpp/__init__.py:20]: Also, I was thinking that we could just run the compilation when the file is not found to avoid manual steps (those will complicate our setup if we want to run experiments remotely also I had to chmod the compile script), I will send a PR to do that later.

Comment (Innixma) [reply autogluon_zeroshot/metrics/_roc_auc_cpp/__init__.py:20]: sounds good!
